OAuth2 Device Flow + Testing Coverage

colemickens opened the pull request:
1. Add support for the OAuth2 Device Flow

2. Adds tests for the Device Flow

3. Adds helpers and tests for persisting and restoring tokens

4. Adds support for manual "bring-your-own-tokens" in ServicePrincipalToken

5. Add tests for manual token and certificate based ServicePrincipalToken

6. Updates the example with support for device auth, and persisting and restoring the token between executions

The code coverage in `autorest/azure` is now:
* azure.go: `96.2%`
* devicetoken.go: `100%`
* persist.go: `92.0%`
* token.go: `96.2%`

colemickens commented:
[coverage.html in a zip](https://github.com/Azure/go-autorest/files/104695/coverage.zip) for those that are curious and/or lazy.

brendandixon commented:
@colemickens Any reason for not taking it to 100%

colemickens commented:
I can add a test for the Azure Error stuff that Ahmet added which will boost it a bit more. Otherwise the other error conditions seemed to me like things that would only occur due to transient IO errors. If you have suggestions on how I can hit those branches, I will implement them.

ahmetalpbalkan commented on autorest/azure/devicetoken.go:
```diff
@@ -0,0 +1,196 @@
+
+// WaitForUserCompletion calls CheckForUserCompletion repeatedly until a token is granted or an error state occurs.
+// This prevents the user from looping and checking against 'ErrDeviceAuthorizationPending'.
+func WaitForUserCompletion(client *autorest.Client, clientID string, code *DeviceCode) (*Token, error) {
```
Does this loop indefinitely? We'll probably need a timeout.

colemickens replied:
Per your other comment and writing a bit of it, I'm not going to implement this for now. One, Azure times out anyway. Two, I don't think it's undesirable to loop forever, if anything, there would be a user configurable timeout that would be passed in and respected. Otherwise the user can just call CheckFor... and do whatever they want.

ahmetalpbalkan commented:
Awesome PR! I will review today.

brendandixon commented:
@colemickens Mocks, I use Mocks. And fake up error situations through them. Good testing covers error cases as well as normal path.

colemickens commented:
@brendandixon I've used mocks to cover a large number of error conditions. I'll stop by and we can chat about the few that remain.

@ahmetalpbalkan Azure will timeout after the ExpiresIn interval returned in the DeviceCode. I'll add code that ensures we stop after that time even if Azure doesn't time out.

ahmetalpbalkan commented:
@colemickens in that case that's not very necessary.

brendandixon commented on autorest/azure/devicetoken.go:
```diff
@@ -0,0 +1,196 @@
+		autorest.WithBaseURL(OAuthDeviceEndpoint),
+		autorest.WithFormData(url.Values{
+			"client_id": []string{clientID},
+			"resource":  []string{resource},
```
Any chance these values must be URL encoded?

brendandixon commented on autorest/azure/devicetoken.go:
```diff
@@ -0,0 +1,196 @@
+			"Request for Device Authorization Code resulted in Status: %d: %s",
+			resp.StatusCode,
+			http.StatusText(resp.StatusCode))
+	}
```
When the Response body get closed if an error is returned? Might there be a case where the Body is present?

brendandixon commented on autorest/azure/devicetoken.go:
```diff
@@ -0,0 +1,196 @@
+	err = decoder.Decode(&code)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to decode Device Authorization Code json")
+	}
```
Any reason for not using Respond and ByUnmarshallingJSON?

brendandixon replied:
And, decoding does not close the Body.

brendandixon commented on autorest/azure/devicetoken.go:
```diff
@@ -0,0 +1,196 @@
+			"Exchanging Device Code for OAuth token resulted in Status: %d: %s",
+			resp.StatusCode,
+			http.StatusText(resp.StatusCode))
+	}
```
Ditto as above regarding possibly open Response.Body.

ahmetalpbalkan replied:
yeah this status code check + body close + decoding is repeated in two methods, we can use response decorators here.

brendandixon commented on autorest/azure/devicetoken.go:
```diff
@@ -0,0 +1,196 @@
+	if err != nil {
+		return nil, fmt.Errorf("Failed to decode Device Token json")
+	}
+
```
Ditto as above regarding using Respond and ByUnmarshallingJSON.

brendandixon commented on autorest/azure/devicetoken.go:
```diff
@@ -0,0 +1,196 @@
+
+	// ErrDeviceSlowDown represents the service telling us we're polling too often during device flow
+	ErrDeviceSlowDown = fmt.Errorf("Error while retrieving OAuth token: Slow Down")
+)
```
Must these be exported?

colemickens replied:
Yes. If the user does not wish to use the automatic looping of WaitForUserCompletion then they need these errors to compare against to know how to proceed.

brendandixon commented on autorest/azure/devicetoken.go:
```diff
@@ -0,0 +1,196 @@
+		switch err {
+		case ErrDeviceSlowDown:
+			waitDuration = waitDuration + (intervalDuration / 2)
+		case ErrDeviceAuthorizationPending:
```
Usually, waits get increasingly long (e.g., exponential back-off). These get longer, but they get longer by smaller amounts. Is that what we want?

ahmetalpbalkan replied:
yeah, is this per oauth2 spec? actually I don't prefer increasing intervals.

colemickens replied:
The spec only says: "The client is polling too quickly and should back off at a reasonable rate."

FWIW, even at an absurd rate, I was not able to get Azure to return this error.

Looking though, the wait duration grows by `(intervalDuration / 2)` and intervalDuration is constant as derived from the server's specified interval rate. So the wait period increasing at a constant rate.

With those in mind do we still want to change this or no?

ahmetalpbalkan replied:
let's keep as is. my bar here is that after a minute, this should not be turning into something that polls every 10+ seconds. Anything in 5-10 secs range is reasonable to me.

brendandixon commented on autorest/azure/devicetoken_test.go:
```diff
@@ -0,0 +1,222 @@
+	code, err := InitiateDeviceAuth(client, TestClientID, TestResource)
+	if err != nil {
+		t.Errorf("unexpected error initiating device auth")
+	}
```
We've been prepending all emitted errors with "autorest: " so they're clear in the output (e.g., see responder_test.go).

brendandixon commented on autorest/azure/devicetoken_test.go:
```diff
@@ -0,0 +1,222 @@
+	if err == nil {
+		t.Errorf("failed to get err") // (expecting Error occurred while requesting Device Authorization Code)
+	}
+}
```
Could you, please, add tests to ensure the Body is properly closed under all conditions?

brendandixon commented on autorest/azure/devicetoken_test.go:
```diff
@@ -0,0 +1,222 @@
+	_, err := InitiateDeviceAuth(client, TestClientID, TestResource)
+	if err == nil {
+		t.Errorf("failed to get error") // (expecting Failed to decode Device Authorization Code json)
+	}
```
Should we test to ensure we get the correct kind of error back?

brendandixon commented on autorest/azure/devicetoken_test.go:
```diff
@@ -0,0 +1,222 @@
+	_, err := InitiateDeviceAuth(client, TestClientID, TestResource)
+	if err == nil {
+		t.Errorf("failed to get err") // (expecting Request for Device Authorization Code resulted in Status)
+	}
```
Should we test to ensure we get the correct kind of error back? Ditto throughout.

ahmetalpbalkan replied:
+1

colemickens replied:
String matching? Or export all the errors?

ahmetalpbalkan replied:
these (a few tests above/below) will be expecting of those declared errors, right? if so `==` should do fine.

brendandixon replied:
Agreed.

colemickens replied:
Whoops, didn't realize it was an exported one. Those should certainly be tested. Sure, there are other error tests that do not refer to exported errors. Do we want assertions on substrings for those? Or just these since they're exported?

colemickens replied:
Actually, I've already used the typed errors in places where they are returned. The only places that just check for non-nil would have to use sub string matching. I was doing it by hand last night, I don't mind writing them.

I just don't know if that's idiomatic or desirable?

brendandixon replied:
At present the test could be pass and the code still incorrect (e.g., the code returns the wrong error). That's the hole I believe we should close.

colemickens replied:
Agree. Looking for a push towards substr matching or exporting more errors.

brendandixon replied:
There's no need to export the errors since test code resides in the same package. However, I'm fine with substr checks as well.

ahmetalpbalkan replied:
@brendandixon I think exporting errors is meant for if someone else were to implement their own WaitForDeviceToken logic, they could use these exported errors in their state machine.

brendandixon commented on autorest/azure/devicetoken_test.go:
```diff
@@ -0,0 +1,222 @@
+		time.Sleep(1 * time.Second)
+		sender.EmitContent(errorDeviceTokenResponse("access_denied"))
+	}()
+
```
Go routines may not run w/o a sleep or some other break to give the Go scheduler a moment. You may need to add that here to ensure a reliable test.

colemickens replied:
ignoring this in favor of Ahmet's solution and removing the sleep entirely.

brendandixon commented on autorest/azure/devicetoken_test.go:
```diff
@@ -0,0 +1,222 @@
+		time.Sleep(1 * time.Second)
+		sender.EmitContent(errorDeviceTokenResponse("access_denied"))
+	}()
+
```
Ditto as above.

colemickens replied:
(ditto as well) "ignoring this in favor of Ahmet's solution and removing the sleep entirely."

brendandixon commented on autorest/azure/devicetoken_test.go:
```diff
@@ -0,0 +1,222 @@
+	}()
+
+	_, _ = WaitForUserCompletion(client, TestClientID, deviceCode())
+}
```
Where is the test in this routine? I see no line ever emitting a t.Errorf.

brendandixon commented on autorest/azure/devicetoken_test.go:
```diff
@@ -0,0 +1,222 @@
+	}
+	if err != ErrDeviceAccessDenied {
+		t.Errorf("got wrong error")
+	}
```
I believe it is enough to assert the correct error is returned. Nil will be, by definition, in that case "not a correct" error. Ditto throughout.

brendandixon commented on autorest/azure/example/main.go:
```diff
@@ -17,83 +18,128 @@ import (
 const resourceGroupURLTemplate = "https://management.azure.com/subscriptions/{subscription-id}/resourcegroups"
 const apiVersion = "2015-01-01"
+const xplatClientID = "04b07795-8ddb-461a-bbee-02f9e1bf7b46"
 
```
Is this some magical, unchanging identity?

colemickens replied:
Yes. It's the xplat client id. I debated whether or not to leave it in. It's makes the example easier to use, but I'm fine with removing it.

brendandixon commented on autorest/azure/persist_test.go:
```diff
@@ -0,0 +1,132 @@
+	if *actualToken != expectedToken {
+		t.Errorf("failed to decode properly actual(%v) expected(%v)", *actualToken, expectedToken)
+	}
+}
```
Pedantic comment: This test tests multiple things.

ahmetalpbalkan replied:
@brendandixon that's idiomatic, I don't see a problem with that. t.Fatal/t.Error difference is exactly for this purpose.

ahmetalpbalkan commented:
I feel like we should move devicetoken into a separate package. Mostly because it's an isolated feature. I'm not sure if it's technically possible as I'm not sure if it is relying on something from ./autorest/azure.

ahmetalpbalkan commented on autorest/azure/devicetoken.go:
```diff
@@ -0,0 +1,196 @@
+	if err != nil {
+		return nil, fmt.Errorf("Error occurred while requesting Device Authorization Code. %s", err)
+	}
+	if resp.StatusCode != http.StatusOK {
```
I think we can use responder decorators here to do status checks etc. (i.e. this package can have its own `withErrorUnlessStatusCode()` and you can close the body and deserialize json in that responder list easily as Brendan said.

ahmetalpbalkan commented on autorest/azure/devicetoken.go:
```diff
@@ -0,0 +1,196 @@
+	Interval        *int64  `json:"interval,string"`
+
+	Message  *string `json:"message"` // Azure specific
+	Resource *string // store this as well, needed during token exchange for azure
```
maybe have a `json:",omitempty"` on this field just to be on the safe side?

ahmetalpbalkan commented on autorest/azure/devicetoken.go:
```diff
@@ -0,0 +1,196 @@
+
+var (
+	// ErrDeviceGeneric represents an unknown error from the token endpoint when using device flow
+	ErrDeviceGeneric = fmt.Errorf("Error while retrieving OAuth token: Unknown Error")
```
we need to prefix all errors with `autorest/azure/devicetoken:` or `devicetoken:`

ahmetalpbalkan commented on autorest/azure/devicetoken_test.go:
```diff
@@ -0,0 +1,222 @@
+	client := &autorest.Client{Sender: sender}
+
+	go func() {
+		time.Sleep(1 * time.Second)
```
I'm usually against sleeps in tests. Instead perhaps you can create a new mock that gets a list of responses (`["authorization_pending", "access_denied"]`) and on the first request, returns the first response, on the 2nd returns the second and so on.

brendandixon replied:
Agreed. Better idea. My main point is that go-routines are not reliable (in a sense).

colemickens replied:
Yes, I much prefer this. Thanks!

ahmetalpbalkan commented on autorest/azure/example/main.go:
```diff
@@ -104,27 +150,92 @@ func getResourceGroups(client *autorest.Client) (*string, error) {
+			spt, err = getSptFromCertificate(applicationID, tenantID, resource, certificatePath)
+		}
+		if err != nil {
+			log.Fatalln("failed to retrieve toke:", err)
```
typo

ahmetalpbalkan commented on autorest/azure/example/main.go:
```diff
@@ -17,83 +18,128 @@ import (
+	return spt, nil
+}
+
+func getSptFromDeviceFlow(clientID, tenantID, resource string) (*azure.ServicePrincipalToken, error) {
```
Hmm the terminology is a bit confusing to me. Even though we do device flow, is what we get back a ServicePrincipalToken or just an oauth2.Token?

colemickens replied:
You get back an SPT. That's how you get autorefreshing, etc. Also, SPT implements the interface needed to be assigned to `autorest.Client`'s `Authorizer` field

colemickens replied:
An SPT is just a convenience wrapper around a token. `getSptFromDeviceFlow` gets a token, and then creates the SPT wrapper and returns it.

ahmetalpbalkan replied:
ok I guess I'm not really familiar with SPT terminology at all. I hope our users are not as ignorant as I am. :)

colemickens commented on autorest/azure/token.go:
```diff
@@ -215,23 +256,25 @@ func (spt *ServicePrincipalToken) Refresh() error {
-	if err != nil {
-		return err
+	if spt.RefreshToken != "" {
+		v.Set("grant_type", OAuthGrantTypeRefreshToken)
```
On one hand, I could have a flag, or do a type assertion. On the other hand, I'd prefer to use refreshtoken anytime it's available and it will always be valid to do so. I lean towards leaving as-is.

ahmetalpbalkan commented on autorest/azure/token.go:
```diff
@@ -168,12 +191,24 @@ func NewServicePrincipalTokenWithSecret(id string, tenantID string, resource str
+}
+
+// NewServicePrincipalTokenFromManualToken creates a ServicePrincipalToken using the supplied token
+func NewServicePrincipalTokenFromManualToken(id string, tenantID string, resource string, token Token) *ServicePrincipalToken {
```
I think this method can take the `refreshCallback` value and store it internally. I'm not sure the `SetRefreshCallback` setter will be called twice in any scenario. Also by making it settable only in the constructor you make it concurrency-safe (in other words, Go assignments are not atomic).

colemickens replied:
That means ~3 extra constructors then. Duplicates of the existing constructors with a suffix that indicates they take a refreshCallback. Or it means breaking the existing constructors.

I agree that the method either needs to be removed or locking added. I'm just not sure which.

ahmetalpbalkan replied:
hmm not sure why you would need new constructors? can't we just extend them with an extra nillable `refreshFunc` argument

colemickens replied:
Those would be breaking. @brendandixon noted that they can be added in a backward compatible member if it is a variadic parameter.

ahmetalpbalkan commented on autorest/azure/example/main.go:
```diff
@@ -104,27 +150,92 @@ func getResourceGroups(client *autorest.Client) (*string, error) {
+}
 
-	return &contentsString, nil
+func saveToken(spt azure.Token) {
```
I think a refreshfunc should return an error. So I think you should define

    type TokenRefreshCallback func (azure.Token) error

and use that in the `SetRefreshCallback()`. I think the error returned from the callback (i.e. failure to save the token to a file for instance) can be handled by the `Refresh()` function in the autorest ––simply because when such a case happens, autorest is the caller and it is the package being executed.

Otherwise your only option in this callback is to a) panic b) log.Fatal/os.Exit(1). Neither are good nor idiomatic.

colemickens replied:
What would you expect Refresh() to do with an error here? does it return it all the way out and make Refresh() error (which will kill off your in-flight request) ?

ahmetalpbalkan replied:
sounds reasonable to me.

ahmetalpbalkan replied:
especially because in the callback I can't handle the error fine because this is all single-threaded code and the code is currently executing the request. so if I won't be able to save the token, the request might as well fail.

colemickens commented:
I've now either replied to, or fixed all issues that have been brought up until now.

[Code coverage is higher](https://s3.amazonaws.com/colemickens-random/devicetoken-iter2.html) (still a few branches that I can't see an easy way to test)

RE: Body Closing: I added some `defer resp.Body.Close()` and then added tests for devicetoken that ensures the response is always closed. I also added another in token_test.go as an extra check that ensures that after `spt.Refresh()` is complete, we've closed the body.

Errors and test errors are prefixed now in accordance with how other packages are doing so.

Only outstanding issues:

1. Is it okay to leave xplat client id in example?

1. Do we split this into a separate package as @ahmetalpbalkan suggested? I think it's fine in here. There's basically no way to use it without the autorest package. While it's not explicitly tied to the autorest/azure package, I can't imagine any time it would *ever* be used without it.

brendandixon commented on autorest/azure/devicetoken.go:
```diff
@@ -0,0 +1,200 @@
+type DeviceToken struct {
+	Token
+	TokenError
+}
```
@colemickens This union-that's-not-a-union bothers me. Is there a benefit in returning the TokenError within the DeviceToken struct? If not, why not define TokenError such that it implements the error interface, then return nil for DeviceToken and the new, improved TokenError as the error?

brendandixon commented on autorest/azure/devicetoken.go:
```diff
@@ -0,0 +1,200 @@
+	if err != nil {
+		return nil, fmt.Errorf("%s %s: %s", logPrefix, errCodeSendingFails, err)
+	}
+	defer resp.Body.Close()
```
The defer rest.Body.Close() here, and anywhere followed by a Responder chain that includes ByClosing, is unnecessary. ByClosing *always* closes the Body. The code need only close the body  *if* it does not invoke a Responder chain that includes a ByClosing. Of course, then I'll ask why it's not using a Responder chain...

colemickens commented:
Removed the defers and made the union struct private.

I was going to add one more commit - adding `v3.1.0` to the Changelog... but I noticed that `autorest/version.go` still says `v1.1.1`. Should I go ahead and update `version.go` to 3.1.0 or is `version.go` for something else?

brendandixon commented:
@colemickens Yes, please update it. Thank you.

colemickens commented:
Version updated. (And fixed a small break that I'd introduced to a public api).

brendandixon commented:
LGTM.

ahmetalpbalkan commented:
LGMT too, we can iterate later. I still couldn't find time to integrate this, I will today.

colemickens commented:
I've added one more commit: https://github.com/colemickens/go-autorest/commit/86b3e80284bc12881a55facbbf65d837496b97b7

We back off with `waitDuration += waitDuration`. If the service tells us to slow down more than 2 times, we will fail and return an error.

@ahmetalpbalkan, @brendandixon, we okay with this?

brendandixon commented:
LGTM
